src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt

```diff
@@ -0,0 +1,89 @@
+import com.arsiu.eduhub.model.Assignment
+import com.arsiu.eduhub.model.Chapter
+import com.arsiu.eduhub.model.Course
+import com.arsiu.eduhub.model.Lesson
+import com.arsiu.eduhub.model.enums.Role
+import com.arsiu.eduhub.service.UserService
+import com.arsiu.eduhub.service.util.mail.EmailParameters
+import com.arsiu.eduhub.service.util.mail.EmailServiceImpl
+import kotlinx.coroutines.DelicateCoroutinesApi
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import org.springframework.beans.factory.config.BeanPostProcessor
+import org.springframework.stereotype.Component
+import java.lang.reflect.InvocationHandler
+import java.lang.reflect.Proxy
+import kotlin.reflect.KClass
+import kotlin.reflect.full.findAnnotation
+import kotlin.reflect.full.memberFunctions
+
+@Component
+@Suppress("SpreadOperator")
+class NotifyOnChangeAnnotationBeanPostProcessor(
+    val userService: UserService,
+    val emailServiceImpl: EmailServiceImpl
+) : BeanPostProcessor {
+
+    val map = mutableMapOf<String, KClass<*>>()
+
+    override fun postProcessBeforeInitialization(bean: Any, beanName: String): Any? {
+        val beanClass = bean::class
+
+        for (method in beanClass.java.methods) {
+            val annotation = method.getAnnotation(NotifyOnChange::class.java)
+            if (annotation != null) {
+                map[beanName] = beanClass
+                break
+            }
+        }
+
+        return bean
+    }
+
+    @OptIn(DelicateCoroutinesApi::class)
+    override fun postProcessAfterInitialization(bean: Any, beanName: String): Any? {
+        val beanClass = map[beanName]
+
+        if (beanClass != null) {
+            return Proxy.newProxyInstance(
+                beanClass.java.classLoader,
+                beanClass.java.interfaces,
+                InvocationHandler { _, method, args ->
+                    val result = method.invoke(bean, *(args ?: emptyArray()))
+
+                    val annotation = beanClass.memberFunctions.firstOrNull {
+                        (it.name == method.name) &&
+                                (it::class.java.typeParameters.contentEquals(method::class.java.typeParameters))
+                    }?.findAnnotation<NotifyOnChange>()
+
+                    if (annotation != null) {
+                        GlobalScope.launch {
+                            val info = when (result) {
+                                is Course -> result.name
+                                is Chapter -> result.course.name
+                                is Lesson -> result.chapter.course.name
+                                is Assignment -> result.lesson.chapter.course.name
+                                else -> ""
+                            }
+                            notifyUsers("${annotation.value} $info")
+                        }
+                    }
+
+                    return@InvocationHandler result
+                }
+            )
+        }
+
+        return bean
+    }
```

Comment:
Have you run detekt?

I believe it reports high code complexity on this function. What about splitting to few smaller and simpler functions?

Reply:
detekt report 
<img width="496" alt="image" src="https://github.com/ArSiu/eduHub/assets/33807305/e92e9b85-c2c8-4654-bc15-754204744daf">